### PR TITLE
mgr/volumes/nfs: Sleep if ganesha disconnects

### PR DESCRIPTION
**Pull request by varshar16**
Fixes: https://tracker.ceph.com/issues/46046
Fixes: https://tracker.ceph.com/issues/46104
Signed-off-by: Varsha Rao <varao@redhat.com>

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
this PR makes #35644 obsolete

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 1]
```diff
@@ -246,14 +245,29 @@ def _gen_export_id(self):
...
+
     def _write_raw_config(self, conf_block, obj, append=False):
         raw_config = GaneshaConfParser.write_conf(conf_block)
         with self.mgr.rados.open_ioctx(self.rados_pool) as ioctx:
             if self.rados_namespace:
                 ioctx.set_namespace(self.rados_namespace)
             if append:
                 ioctx.append(obj, raw_config.encode('utf-8'))
-                ioctx.notify(obj)
+                self._check_rados_notify(ioctx, obj)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/volumes/fs/nfs.py:270]: is 

https://github.com/ceph/ceph/blob/07358ef9045efa65061a0ebb354c1f4093f09d8e/src/pybind/mgr/cephadm/services/nfs.py#L160

also affected by this?

Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py:270]: No, the object gets created successfully. Since ganesha is failing, I hope this will help to check ganesha logs: https://github.com/varshar16/ceph/commit/e016f5a0640e7e60e1b6eb6461b54e8ea6c3bc5e

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 2]
```diff
@@ -246,14 +245,29 @@ def _gen_export_id(self):
...
+            try:
+                ioctx.notify(obj)
+                break
+            except Exception as e:
+                if e.errno == e.ETIMEDOUT:
+                    sleep_time += 10
+                    log.exception(f"Ganesha timed out {e}")
+                    time.sleep(sleep_time)
+                else:
+                    raise
```
Comment (jtlayton) [inline src/pybind/mgr/volumes/fs/nfs.py:260]: Do we really need to retry here? If ganesha dies, then when it comes back it'll pick up the latest config. At that point the notification is not really necessary anymore.

Comment (batrick) [reply src/pybind/mgr/volumes/fs/nfs.py:260]: I agree with Jeff. I don't think we need to do a retry. We should just ignore `ETIMEDOUT`.

Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py]: @batrick @jtlayton I have updated the patch. Please take another look.

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 3]
```diff
@@ -246,14 +244,23 @@ def _gen_export_id(self):
                 break
         return nid
 
+    def _check_rados_notify(self, ioctx, obj):
```
Comment (ajarr) [inline src/pybind/mgr/volumes/fs/nfs.py:247]: this can be a static method?

Comment (varshar16) [reply src/pybind/mgr/volumes/fs/nfs.py]: Changed

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 4]
```diff
@@ -3,8 +3,6 @@
 import logging
 
 from ceph.deployment.service_spec import NFSServiceSpec, PlacementSpec
-
-import cephfs
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/nfs.py:7]: This should be a separate commit.

File: qa/tasks/cephfs/test_nfs.py [hunk 5]
```diff
@@ -64,9 +64,15 @@ def _test_create_cluster(self):
...
-        self.assertEqual("No services reported\n", orch_output)
+        expected_output = "No services reported\n"
+        wait_time = 10
+        while wait_time <= 60:
+            time.sleep(wait_time)
+            orch_output = self._check_nfs_status()
+            if expected_output == orch_output:
+                return
+            wait_time += 10
+        raise RuntimeError("NFS Ganesha cluster could not be deleted")
```
Comment (batrick) [inline qa/tasks/cephfs/test_nfs.py:75]: `self.fail`

File: src/pybind/mgr/volumes/fs/nfs.py [hunk 6]
```diff
@@ -198,6 +196,16 @@ def __init__(self, mgr, namespace=None):
         self.export_conf_blocks = []
         self.exports = {}
 
+    @staticmethod
+    def _check_rados_notify(ioctx, obj):
+        try:
+            ioctx.notify(obj)
+        except Exception as e:
```
Comment (batrick) [inline src/pybind/mgr/volumes/fs/nfs.py:203]: I think we can just catch `rados.TimedOut`?

**Comment by varshar16** [comment]
Tests passed: http://qa-proxy.ceph.com/teuthology/varsha-2020-06-22_04:37:56-rados-wip-fix-rados-ganesha-timeout-distro-basic-smithi/

**Comment by varshar16** [comment]
Removed commit c4c5bb6 as Kefu has a patch for it. (#35674)

**Comment by varshar16** [comment]
http://qa-proxy.ceph.com/teuthology/varsha-2020-06-23_12:55:10-rados-wip-fix-rados-ganesha-timeout-distro-basic-smithi/
